components/autofill/src/db/store.rs

```diff
@@ -2,178 +2,60 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-use crate::db::models::address::{Address, UpdatableAddressFields};
-use crate::db::models::credit_card::{CreditCard, UpdatableCreditCardFields};
+use crate::db::models::address::{Address, InternalAddress, UpdatableAddressFields};
+use crate::db::models::credit_card::{CreditCard, InternalCreditCard, UpdatableCreditCardFields};
 use crate::db::{addresses, credit_cards, AutofillDb};
 use crate::error::*;
+use crate::sync::engine::ConfigSyncEngine;
 use rusqlite::{
     types::{FromSql, ToSql},
     Connection,
 };
 use sql_support::{self, ConnExt};
-use std::cell::RefCell;
 use std::path::Path;
-use std::sync::{Arc, Mutex, Weak};
-use sync15_traits::SyncEngine;
+use std::sync::{Arc, Mutex};
 use sync_guid::Guid;
 
 // Our "sync manager" will use whatever is stashed here.
 lazy_static::lazy_static! {
     // Mutex: just taken long enough to update the inner stuff - needed
-    //        to wrap the RefCell as they aren't `Sync`
-    // RefCell: So we can replace what it holds. Normally you'd use `get_ref()`
-    //          on the mutex and avoid the RefCell entirely, but that requires
-    //          the mutex to be declared as `mut` which is apparently
-    //          impossible in a `lazy_static`
-    // [Arc/Weak]<StoreImpl>: What the sync manager actually needs.
-    pub static ref STORE_FOR_MANAGER: Mutex<RefCell<Weak<StoreImpl>>> = Mutex::new(RefCell::new(Weak::new()));
+    //        to wrap the Option<Store> as they aren't `Sync`
+    static ref STORE_FOR_MANAGER: Mutex<Option<Store>> = Mutex::new(None);
 }
 
-// This is the type that uniffi exposes. It holds an `Arc<>` around the
-// actual implementation, because we need to hand a clone of this `Arc<>` to
-// the sync manager and to sync engines. One day
-// https://github.com/mozilla/uniffi-rs/issues/417 will give us access to the
-// `Arc<>` uniffi owns, which means we can drop this entirely (ie, `Store` and
-// `StoreImpl` could be re-unified)
-// Sadly, this is `pub` because our `autofill-utils` example uses it.
+// Get a Store for the SyncManager to use
+pub fn get_store_for_manager() -> Option<Store> {
+    STORE_FOR_MANAGER.lock().unwrap().as_ref().cloned()
+}
+
+// This is the type that uniffi exposes. It has `Arc<>` around the db mutex because
+// register_with_sync_manager() needs to clone ourself.  This is redundant though, because uniffi
+// also has an Arc<>.  One day https://github.com/mozilla/uniffi-rs/issues/419 will give us access
```

Comment:
This piece of the UniFFI `Arc` work is ready on `main` as of yesterday, I think I'll go ahead and get it into a release so we can try it out.

Reply:
If you update to UniFFI `v0.10.0` (like [this](https://github.com/mozilla/application-services/pull/4131) then you should be able to use the new `[Self=ByArc]` annotation to directly obtain the underlying `Arc`.